definitions/npm/firebase_v5.x.x/flow_v0.34.x-/firebase_v5.x.x.js

```diff
@@ -541,24 +544,28 @@ declare interface $npm$firebase$firestore$DocumentChange {
 declare class $npm$firebase$firestore$DocumentReference {
   firestore: $npm$firebase$firestore$Firestore;
   id: string;
+  path: string;
   parent: typeof $npm$firebase$firestore$CollectionReference;
   collection(collectionPath: string): $npm$firebase$firestore$CollectionReference;
   delete(): Promise<void>;
   get(): Promise<$npm$firebase$firestore$DocumentSnapshot>;
+  isEqual(other: $npm$firebase$firestore$DocumentReference): boolean;
   onSnapshot(
-    optionsOrObserverOrOnNext: $npm$firebase$firestore$QueryListenOptions | $npm$firebase$firestore$documentObserver,
+    optionsOrObserverOrOnNext: $npm$firebase$firestore$QueryListenOptions
+    | $npm$firebase$firestore$documentObserver,
     observerOrOnNextOrOnError?: | $npm$firebase$firestore$QueryListenOptions
     | $npm$firebase$firestore$documentObserver
     | $npm$firebase$firestore$observerError,
     onError?: $npm$firebase$firestore$observerError
-  ): Function;
+  ): () => void;
   set(data: Object, options?: {| merge?: boolean, mergeFields?: string[] |} | null): Promise<void>;
   update(...args: Array<any>): Promise<void>;
 }
 
 declare class $npm$firebase$firestore$DocumentSnapshot {
-  data(): Object;
+  data(): Object | void;
```

Comment:
Would be nice to replace `Object` while at it, cause it's an alias of `any` from Flow >= 0.75.  
`{ +[string]: mixed }` ?

Reply:
Although it is more sound to declare this as `{ +[string]: mixed }`, it might be too noisy for existing codebases.
If you see the _Listen to multiple documents in a collection_ in [official docs](https://firebase.google.com/docs/firestore/query-data/listen), the usage of the `data()` method is similar to that of `JSON.parse`.

Here is a example to show that it might be too noisy for existing codebases:
https://flow.org/try/#0PQKgBAAgZgNg9gdzAZwC4CcCWBjVYTABQhqAngA4CmYAIgIap1gC8YA3mANQDaaWAdgHMAugC4wAW0wAPSgBMwAXwDcxKAFd+uTHH5hBlVPUYAKAJTjjTNoTBhUAC3SIw-SkgCi6Z+nPKwhIrEZFRgAAqU6Mi6LOwAPrZc-HQSlOJ8mEKqdpx0BuL86hIARpGqcSrE2Lpo7K4plAA0YHnUimDiEVExrAZGDHR+hEA

I'm also thinking of remove the newly added `void` case because there are many places where you definitely know that the data exist. The example shown in the official docs is one such instance.

My proposed solution is `{| +[string]: any |}`. I use exact object notation so that it can be compatible with exact type declaration. Here is the demo:
https://flow.org/try/#0PTAEAEDMBsHsHdQGcAuAnAlgYxQKFygJ4AOApqACICGKVoAvKAN4A+oA1ANqqYB2A5gF0AXKCq9CoFgF8A3PkgBXXjgyxeofqRTVaACgCUo3XSa5QoFAAs0CUL1KIAomltpDs0Lmn4iZUAAKpGhI6gzMLOYcvFQAtqSiPBgC8hbsVFqivIqxAEbB8jLyuFjqqMz2caQANGJaoNKiQSFhjFo6NFQeuEA

If you like it, I can quickly make the change.

Reply:
Your right `mixed` is more suited for inputs than return types. 
`{| +[string]: any |}` seems reasonable in such case.  

Seems to fit their TS declarations also:
https://github.com/firebase/firebase-js-sdk/blob/master/packages/firebase/index.d.ts#L1556
https://github.com/firebase/firebase-js-sdk/blob/master/packages/firebase/index.d.ts#L754

Reply:
I also converted those unclear object types in input positions to `{ +[string]: mixed }`. I think it's good for merging now.